File: 240327/네 방향 탈출 가능 여부 판별하기/determine-escapableness-with-4-ways.cpp
// https://www.codetree.ai/missions/2/problems/determine-escapableness-with-4-ways/description
#include <iostream>
#include <queue>
using namespace std;

int n, m;
int dx[4] = { 1,0,-1,0 };
int dy[4] = { 0,1,0,-1 };
int grid[101][101];			// 인접 행렬 기반 그래프
bool visited[101][101]; 
queue<pair<int, int>> q;	// 노드 = 좌표 = pair<int, int>

bool InRange(int x, int y) {
	return x >= 0 && x < n && y >= 0 && y < m;
}

bool canGo(int x, int y) {
	if (!InRange(x, y))
		return false;
	else if (visited[x][y] || grid[x][y] == 0)
		return false;
	else
		return true;
}

int main() {
	cin >> n >> m;

	for (int i = 0; i < n; i++) {
		for (int j = 0; j < m; j++) {
			cin >> grid[i][j];
		}
	}

	q.push(make_pair(0, 0));
	visited[0][0] = true;

	// BFS 수행 (반복문)
	while (!q.empty()) {
		int cur_x = q.front().first;
		int cur_y = q.front().second;
		q.pop();

		for (int i = 0; i < 4; i++) {
			int new_x = cur_x + dx[i];
			int new_y = cur_y + dy[i];

			if (canGo(new_x, new_y)) {
				visited[new_x][new_y] = true;
				q.push(make_pair(new_x, new_y));
			}
		}
	}

	if (visited[n - 1][m - 1])
		cout << 1 << '\n';
	else
		cout << 0 << '\n';

	return 0;
}

File: 240629/네 방향 탈출 가능 여부 판별하기/determine-escapableness-with-4-ways.js
// 문제 : https://www.codetree.ai/missions/2/problems/determine-escapableness-with-4-ways/description

/*
    뱀이 있는 경우 : 0
    뱀이 없는 경우 : 1

    뱀한테 물리지 않고 탈출할 수 있는 경우 ? 1 : 0 출력

    BFS 유형 -> DFS로도 풀 수 있을 듯
*/

let N = null;
let M = null;
let countN = 0;

let graph = [];
const dx = [-1, 0, 1, 0];
const dy = [0, 1, 0, -1];

class Queue {
    constructor() {
        this._arr = []; // 값을 저장할 객체
    }

    enqueue(item) {
        this._arr.push(item);
    }

    dequeue() {
        return this._arr.shift();
    }

    isEmpty() {
        return this._arr.length === 0 ? true : false;
    }
}

const solutionByBFS = (graph, visited) => {
    const inRange = (x, y) => {
        return x >= 0 && x < N && y >= 0 && y < M;
    }

    const canGo = (x, y) => {
        if (!inRange(x, y)) return false;
        else if (visited[x][y] || !graph[x][y]) return false;
        else return true;
    }
    
    const queue = new Queue();

    queue.enqueue([0, 0]);
    visited[0][0] = true;

    while (!queue.isEmpty()) {
        let [curX, curY] = queue.dequeue();

        for (let i = 0; i < 4; i++) {
            let newX = curX + dx[i];
            let newY = curY + dy[i];

            if (canGo(newX, newY)) {
                queue.enqueue([newX, newY]);
                visited[newX][newY] = true;
            }
        }
    }

    if (visited[N - 1][M - 1]) {
        console.log(1);
    }
    else {
        console.log(0);
    }
}

const solutionByDFS = (graph, visited) => {
    const inRange = (x, y) => {
        return x >= 0 && x < N && y >= 0 && y < M;
    }

    const canGo = (x, y) => {
        if (!inRange(x, y)) return false;
        else if (visited[x][y] || !graph[x][y]) return false;
        else return true;
    }

    const dfs = (x, y) => {
        visited[x][y] = true;

        for (let i = 0; i < 4; i++) {
            let newX = x + dx[i];
            let newY = y + dy[i];

            if (canGo(newX, newY)) {
                dfs(newX, newY);
            }
        }
    }

    dfs(0, 0);

    if (visited[N - 1][M - 1]) {
        console.log(1);
    }
    else {
        console.log(0);
    }
}

const rl = require('readline').createInterface({
    input: process.stdin,
    output: process.stdout,
});

rl.on('line', (line) => {
    if (!N && !M) {
        [N, M] = line.split(' ').map(e => +e);
    }
    else {
        graph.push(line.split(' ').map(e => +e));
        countN++;
    }

    if (countN === N) {
        rl.close();
    }
}).on('close', () => {
    let visited = Array.from(Array(N), () => Array(M).fill(false));
    // solutionByDFS(graph, visited);
    solutionByBFS(graph, visited);
    process.exit();
});